### ISSUE TITLE
Code Review 2

### ISSUE THREAD

**Issue by cdmckinley** [eemclaughlin/indieProject#9]
# Design/Code Review 2 @pawaitemadisoncollege

#### Reviewer: Craig McKinley

|Areas for Improvement|Criteria|Items Met or Exceeded|
|--------|---------|---------------|
||**Project effectively utililizes the technologies and technques specified in the project objectives**||
|| Planned MVP Functionality | Seems sound! I like the idea of the database containing references to pages of cookbooks, instead of storing redundant, likely inaccurate data. |
|Should not print Stack Trace in exceptions. I'd recommend `logger.error(exception)`| Logging framework used, i.e., no System.out.println() or printStacktrace() statements.| Makes much use of logging using Log4J, with no `System.out.println` statements |
|| Hibernate used for all data access. | Makes use of hibernate for entity data |
|| Authentication implemented. | Makes use of Cognito for Auth |
|| Consumes at least one web service or public api using Java. | Makes use of Google Books API |
|| Application is database-driven using full CRUD.| Has add, delete, edit, and list controllers |
| I don't know that having the direct relationship for 1 user to many recipes is necessary, as those exist for 1 user to many cookbooks, and 1 cookbook to many recipes. Maybe for V2: I wonder if more of these should be ManyToMany, such as when two users have the same cookbook, with users still having multiple cookbooks. | Database includes multiple one to many relationships. | From One-Many: User-CookBook, User-Recipe, CookBook-Recipe, Recipe-Tag, Tag-Recipe |
|| Deployed to AWS for public access. | It's deployed on AWS! |
| Server-side data validation may still be needed. | Implements best practices (for example, data validation)  | Client side validation loks great! Most sensitive data is kept out of Git. The only exception I see is log4j's properties, which only reveal the user directory. |
|| **Synthesis of multiple concepts in unfamiliar situations requiring research beyond the scope of the class**||
||Experiments individually, exhibits independence and drive, shows originality in the solution.| Storing references to existing data (the cookbooks) is original, and seems very smart! |
||Implemented technologies or techniques not covered in course materials. | Filters data from API that provides much unneeded data |
||**Code quality - Evaluate  code quality for the following and identify specific areas for improvement (class, method or line number). Be sure to list which code quality plugins/tools you used to assist with this analysis**||
|| Single-purpose methods ||
|| Well-structured project||
|| Descriptive naming of packages, classes, methods, variables ||
||Classes appropriately-sized  (no monster classes)| Classes seem appropriately sized. |
|| CPD (copy paste detection, meaning are the same lines of code repeated?| I don't see any warnings for this |
|| Are there candidates for super/subclass relationships, abstract classes, interfaces | Not that I'm seeing |
|| Are any values hard-coded that should be in a properties file? | I'm not noticing any literal values that should be in a properties file |
||Proper exception handling | As far as I can tell |
| It would be great to make an error controller that forwards to an error JSP. The web.xml file can assign redirects per error code. |Proper error reporting to the user - custom erro pages?|| 
||Code documentation| Has very complete JavaDocs, and a directory dedicated to JavaDoc documentation. | 
| Ideally the data validation in the controllers should likely be their own method(s).  | Is there code in the servlet doGet/doPost that should be refactored into testable classes or methods? ||
| I'm not sure on this one. |Evaluate the JSPs for templating, data validation, overall look and feel.| I'm not sure on this one. |
| Minimal amount. Shouldn't be too hard to remove | JSPs use JSTL and EL, no java code | Pretty good overall || |
|| Unit tests are truly a unit test rather than a high level functional test | These test look very good to me. | 
||Test data is appropriately cleaned up or handled| I'm not noticing any data that should be cleaned up and isn't. I'll say it's good! | 
| I believe a test could be useful for the conversion of the data from form inputs to the app data, and from app data to visible data |There is full coverage of methods that perform business logic or utility functions| DAOs seem to be adequately tested. |
||Redundant code is eliminated by using set up and tear down methods, i.e., @Before, @After| The unit test setup uses @BeforeEach appropriately |
| Storing the old JSPs somewhere other than your repository may be a good idea. |Other comments/notes?| I'm impressed with the work done here. Well done! |
||**Demonstrates initiative and thoughtful planning to leverage available resources (time, equipment, external expertise) to meet milestones and project objectives.** ||
|| **Evidence of significant revision and incorporation of feedback.** | Has incorporated some feedback, while seemingly keeping scope and priorities in mind. |
|| **Project complexity** | I'd say at least complex enough. Try not to overwork though |
|| **Additional Comments** ||

**Comment by pawaitemadisoncollege** [comment]
Hi @cdmckinley, thanks for your review.  Three questions for you, 

1. Has all planned mvp functionality been met or are there items left to complete yet? These should be listed in "Planned MVP Functionality" row. 
2. I'm curious about your comment "I'm not sure on this one". Can you share what you aren't sure of? Is there some information you need yet to evaluate this? 
3. Regarding "Demonstrates initiative and thoughtful planning to " - I don't see an evaluation here. How does plan execution look?

**Comment by cdmckinley** [comment]
@pawaitemadisoncollege I'll see if I can offer more complete answers for those, though I want to preface this by letting @eemclaughlin know:
I had trouble with signup. I created an account on Cognito, but something isn't happening correctly. I think it's with the redirect. I got a security warning in the browser, and I think the redirect is trying to send me to HTTPS, which redirects me to HTTP?
Long story short, I'm presently unable to fully demo the application until this is fixed, but I'll try and offer more complete feedback for how an existing user may work with it, based on the home page and project code.

For Planned MVP Functionality: If this is what's outlined in the README, a good amount of this seems to be in order based on the demonstration! I wonder about "Users will also be able to search using recipe name or keywords. When a recipe is found, it will then point the user to the appropriate cookbook and page number." as I can't recall if the page numbers were shown, and I know the tags were still being worked on when Eric presented.

For the JSPs, there are includes for `taglib`, head, navbar, footer, and JavaScript. I wonder if templating could help further with how many JSPs there are? There are many JSPs, so I wonder if they could be condensed a bit. There is client-side data validation as well, and the pages are styled nicely!

For "Demonstrates initiative and thoughtful planning to leverage available resources": It's clear that there was much planning and initiative, as the project is pretty big and must have taken much time! I do wonder what led the project to having so many controllers and JSPs, though. It could have probably been completed with less, but maybe there's a reason that I'm unaware of.

I want to reiterate, well done! I can see a lot has gone into this project, and look forward to seeing the final result
